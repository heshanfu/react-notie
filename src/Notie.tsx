import * as React from 'react';
import * as ReactDOM from 'react-dom';

import {
    NotieActions,
    NotieSettings,
    NotieLevel,
    alert,
    NotieCallback,
    NotieEventDetails
} from './actions';

import {
    NotieContainer,
    NotieMessage,
    NotieChoices,
    NotieButton,
    NotieOverlay
} from './components';

const DEFAULT_STATE = {
    visible: false,
    locked: false,
    position: 'top',
    okBtnText: 'OK',
    cancelBtnText: 'Cancel'
};

export interface NotieProps {}

export interface NotieState extends Pick<NotieSettings, 'message' | 'level'> {
    visible: boolean;
    locked: boolean;
    okBtnText: string;
    cancelBtnText: string;
    position: 'top' | 'bottom';
}

let alreadyLoaded = false;

const DEFAULT_TIMEOUT = 5000;

export class Notie extends React.Component<NotieProps, NotieState> {
    private root: HTMLDivElement;
    private notieRoot: HTMLDivElement | null;
    private transitionendCallback: null | NotieCallback;
    private timeout: number | null;
    private confirmResolve: NotieCallback;
    private confirmReject: NotieCallback;

    constructor(props: NotieProps) {
        super(props);

        this.state = {
            message: '',
            level: NotieLevel.INFO,
            visible: false,
            locked: false,
            position: 'top',
            okBtnText: 'OK',
            cancelBtnText: 'Cancel'
        };
        this.root = document.createElement('div');
    }

    componentWillMount() {
        if (alreadyLoaded)
            throw new Error(
                '<Notie /> is already loaded somwhere in your code. Please include it only once.'
            );

        alreadyLoaded = true;
    }

    componentDidMount() {
        if (this.notieRoot) this.notieRoot.style.opacity = '0';
        window.addEventListener(NotieActions.SHOW, this.showNotie);
        document.body.appendChild(this.root);
    }

    shouldComponentUpdate(_: NotieProps, nextState: NotieState) {
        return this.state.visible !== nextState.visible;
    }

    componentWillUnmount() {
        if (this.transitionendCallback) {
            this.notieRoot &&
                this.notieRoot.removeEventListener('transitionend', this.transitionendCallback);
            this.transitionendCallback = null;
        }

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (alreadyLoaded) alreadyLoaded = false;

        window.removeEventListener(NotieActions.SHOW, this.showNotie);
        document.body.removeChild(this.root);
    }

    showNotie = (event: Event) => {
        if (this.state.locked) return;

        const settings: NotieEventDetails = (event as CustomEvent).detail;
        const toLock = settings.level === NotieLevel.CONFIRM || settings.level === NotieLevel.FORCE;

        if (settings.confirmReject && settings.confirmResolve) {
            this.confirmReject = settings.confirmReject;
            this.confirmResolve = settings.confirmResolve;
        } else {
            delete this.confirmReject;
            delete this.confirmResolve;
        }

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (this.state.visible) {
            this.hideNotie(() => alert(settings));
            return;
        }

        if (this.notieRoot) this.notieRoot.style.opacity = '1';

        this.setState(state => ({
            message: settings.message,
            level: settings.level,
            visible: true,
            locked: toLock,
            okBtnText: settings.yesBtnText || state.okBtnText,
            cancelBtnText: settings.noBtnText || state.cancelBtnText
        }));

        if (!toLock) {
            this.timeout = window.setTimeout(() => {
                this.hideNotie();
                this.timeout = null;
            }, settings.ttl || DEFAULT_TIMEOUT);
        }
    };

    hideNotie = (callback?: NotieCallback) => {
        this.setState(DEFAULT_STATE as NotieState);
        this.transitionendCallback = () => {
            if (this.notieRoot) this.notieRoot.style.opacity = '0';
            if (typeof callback === 'function') callback();
            this.notieRoot &&
                this.notieRoot.removeEventListener('transitionend', this
                    .transitionendCallback as NotieCallback);
            this.transitionendCallback = null;
        };
        this.notieRoot &&
            this.notieRoot.addEventListener('transitionend', this.transitionendCallback);
    };

    handleYes = () => this.hideNotie(this.confirmResolve);

    handleNo = () => this.hideNotie(this.confirmReject);

    handleDismiss = () => {
        if (this.state.locked) return;

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (this.state.visible) {
            this.hideNotie();
        }
    };

    rootRef = (c: HTMLDivElement) => (this.notieRoot = c);

    render() {
        const { message, level, visible, position } = this.state;
        const forced = level === NotieLevel.CONFIRM || level === NotieLevel.FORCE;

        return ReactDOM.createPortal(
            <div ref={this.rootRef}>
                {forced && visible && <NotieOverlay />}
                <NotieContainer
                    active={visible}
                    position={position}
                    level={level}
                    onClick={this.handleDismiss}>
                    <NotieMessage>{message}</NotieMessage>
                    {forced && (
                        <NotieChoices>
                            <NotieButton
                                level={
                                    level === NotieLevel.FORCE
                                        ? NotieLevel.ERROR
                                        : NotieLevel.SUCCESS
                                }
                                onClick={this.handleYes}>
                                {this.state.okBtnText}
                            </NotieButton>
                            {level === NotieLevel.CONFIRM && (
                                <NotieButton level={NotieLevel.ERROR} onClick={this.handleNo}>
                                    {this.state.cancelBtnText}
                                </NotieButton>
                            )}
                        </NotieChoices>
                    )}
                </NotieContainer>
            </div>,
            this.root
        );
    }
}
