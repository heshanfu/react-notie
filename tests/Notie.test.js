import React from 'react';
import { mount } from 'enzyme';

import Notie from '../src/Notie';

jest.useFakeTimers();

describe('<Notie />', () => {
    let component;

    beforeEach(() => {
        component = mount(
            <Notie />
        );
    });

    afterEach(() => {
        if (component) {
            component.unmount();
            component = null;
        }
    });

    test('initial render', () => {
        expect(component).toMatchSnapshot();
    });

    test('success message', () => {
        component.instance().success('Success!');
        component.update();
        expect(component.find('.react-notie-level--success').length).toBe(1);
        expect(component.find('.react-notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.react-notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('warn message', () => {
        component.instance().warn('Warn!');
        component.update();
        expect(component.find('.react-notie-level--warn').length).toBe(1);
        expect(component.find('.react-notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.react-notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('error message', () => {
        component.instance().error('Error!');
        component.update();
        expect(component.find('.react-notie-level--error').length).toBe(1);
        expect(component.find('.react-notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.react-notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('info message', () => {
        component.instance().info('Info!');
        component.update();
        expect(component.find('.react-notie-level--info').length).toBe(1);
        expect(component.find('.react-notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.react-notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('consecutive calls', () => {
        const instance = component.instance();
        instance.info('Info!');
        component.update();
        expect(component.find('.react-notie-level--info').length).toBe(1);
        expect(component.find('.react-notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(2000);
        component.update();
        instance.error('Error!');

        /*
            Must be triggered manually since `transitionend` event
            is not being trigged when using `dispatchEvent` on root
            element.
        */
        instance.transitionendCallback();
        expect(instance.transitionendCallback).toBe(null);
        component.update();

        expect(component.find('.react-notie-level--error').length).toBe(1);
        expect(component.find('.react-notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
    });

    test('cleanup on unmount', () => {
        const instance = component.instance();
        instance.info('Info!');
        instance.info('Info!');
        jest.runTimersToTime(2000);
        component.update();
        component.unmount();
        expect(instance.transitionendCallback).toBe(null);
    });

    test('dismiss click', () => {
        component.instance().info('Info!');
        component.update();
        expect(component.find('.react-notie-level--info').length).toBe(1);
        expect(component.find('.react-notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        component.find('.react-notie-dismiss').simulate('click');
        component.update();
        expect(component.find('.react-notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    describe('confirm', () => {
        test('initial render', () => {
            component.instance().confirm('Are you sure?');
            component.update();
            expect(component.find('.react-notie-level--confirm').length).toBe(1);
            expect(component.find('.react-notie--active').length).toBe(1);
            expect(component).toMatchSnapshot();
        });

        test('click yes', (done) => {
            const instance = component.instance();
            const promise = instance.confirm('Are you sure?');

            component.update();
            component.find('.react-notie-choice--yes').simulate('click');
            instance.transitionendCallback();
            expect(instance.transitionendCallback).toBe(null);
            component.update();

            promise.then(() => {
                expect(component).toMatchSnapshot();
                expect(component.find('.react-notie--active').length).toBe(0);
                done();
            });
        });

        test('click no', (done) => {
            const instance = component.instance();
            const promise = instance.confirm('Are you sure?');

            component.update();
            component.find('.react-notie-choice--no').simulate('click');
            instance.transitionendCallback();
            expect(instance.transitionendCallback).toBe(null);
            component.update();
            promise.catch(() => {
                expect(component).toMatchSnapshot();
                expect(component.find('.react-notie--active').length).toBe(0);
                done();
            });
        });

        test('no consecutive calls', () => {
            const instance = component.instance();
            instance.confirm('Are you sure?');
            component.update();
            expect(component.find('.react-notie-level--confirm').length).toBe(1);
            expect(component).toMatchSnapshot();
            instance.info('Info!');
            component.update();
            expect(component.find('.react-notie-level--confirm').length).toBe(1);
            expect(component).toMatchSnapshot();
        });

        test('Custom button text', () => {
            const instance = component.instance();
            instance.confirm('Are you sure?', {
                yesBtnText: 'Hell Yeah!',
                noBtnText: 'Fuck NO!'
            });
            expect(component).toMatchSnapshot();
        });
    });
});
