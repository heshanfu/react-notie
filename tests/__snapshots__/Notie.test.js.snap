// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notie /> confirm Custom button text 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--info"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      />
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> confirm click no 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--confirm"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Are you sure?
      </div>
      <div
        className="react-notie-choices"
      >
        <div
          className="react-notie-choice react-notie-choice--yes"
          onClick={[Function]}
        >
          Yes
        </div>
        <div
          className="react-notie-choice react-notie-choice--no"
          onClick={[Function]}
        >
          No
        </div>
      </div>
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> confirm click yes 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--confirm"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Are you sure?
      </div>
      <div
        className="react-notie-choices"
      >
        <div
          className="react-notie-choice react-notie-choice--yes"
          onClick={[Function]}
        >
          Yes
        </div>
        <div
          className="react-notie-choice react-notie-choice--no"
          onClick={[Function]}
        >
          No
        </div>
      </div>
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> confirm initial render 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--confirm react-notie--active"
  >
    <div
      className="react-notie-overlay"
    />
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Are you sure?
      </div>
      <div
        className="react-notie-choices"
      >
        <div
          className="react-notie-choice react-notie-choice--yes"
          onClick={[Function]}
        >
          Yes
        </div>
        <div
          className="react-notie-choice react-notie-choice--no"
          onClick={[Function]}
        >
          No
        </div>
      </div>
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> confirm no consecutive calls 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--confirm react-notie--active"
  >
    <div
      className="react-notie-overlay"
    />
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Are you sure?
      </div>
      <div
        className="react-notie-choices"
      >
        <div
          className="react-notie-choice react-notie-choice--yes"
          onClick={[Function]}
        >
          Yes
        </div>
        <div
          className="react-notie-choice react-notie-choice--no"
          onClick={[Function]}
        >
          No
        </div>
      </div>
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> confirm no consecutive calls 2`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--confirm react-notie--active"
  >
    <div
      className="react-notie-overlay"
    />
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Are you sure?
      </div>
      <div
        className="react-notie-choices"
      >
        <div
          className="react-notie-choice react-notie-choice--yes"
          onClick={[Function]}
        >
          Yes
        </div>
        <div
          className="react-notie-choice react-notie-choice--no"
          onClick={[Function]}
        >
          No
        </div>
      </div>
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> consecutive calls 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--info react-notie--active"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Info!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> consecutive calls 2`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--error react-notie--active"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Error!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> dismiss click 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--info react-notie--active"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Info!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> dismiss click 2`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--info"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Info!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> error message 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--error react-notie--active"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Error!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> error message 2`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--error"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Error!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> info message 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--info react-notie--active"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Info!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> info message 2`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--info"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Info!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> initial render 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--info"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      />
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> success message 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--success react-notie--active"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Success!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> success message 2`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--success"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Success!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> warn message 1`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--warn react-notie--active"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Warn!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;

exports[`<Notie /> warn message 2`] = `
<Notie
  position="top"
  ttl={5000}
>
  <div
    className="react-notie react-notie--top react-notie-level--warn"
  >
    <div
      className="react-notie-container"
    >
      <div
        className="react-notie-message"
      >
        Warn!
      </div>
      <div
        className="react-notie-dismiss"
        onClick={[Function]}
      />
    </div>
  </div>
</Notie>
`;
